### PR TITLE
Add --exclude option to delocate-wheel

### PR DESCRIPTION
**Pull request by rossant**
Same motivation as https://github.com/pypa/auditwheel/pull/310

I need this to remove libvulkan from the repaired wheel.

### CODE REVIEW

File: delocate/cmd/delocate_wheel.py [hunk 1]
```diff
@@ -59,6 +69,15 @@ def main():
...
         require_archs = opts.require_archs
     lib_filt_func = 'dylibs-only' if opts.dylibs_only else None
+
+    exclude = [
+        _.strip() for _ in (opts.exclude or '').split(',') if _.strip()]
+
+    def copy_filt(name):
+        if not exclude:
+            return filter_system_libs(name)
+        return filter_system_libs(name) and not _is_in_list(name, exclude)
```
Comment (matthew-brett) [inline delocate/cmd/delocate_wheel.py:79]: I think you are doing a substring search here?  Is that what you intended?  Or did you mean

```
and not name in exclude
```

?

File: delocate/cmd/delocate_wheel.py [hunk 2]
```diff
@@ -37,6 +45,8 @@ def main():
                action="store_true",
                help="Only analyze files with known dynamic library "
                "extensions"),
+        Option("-e", "--exclude",
```
Comment (matthew-brett) [inline delocate/cmd/delocate_wheel.py:48]: Or allow multiple `--exclude` options with `action="append"`?

**Review by matthew-brett — COMMENTED** [review]
Sorry for the delay.  Some thoughts.   Can you think of a test?

File: delocate/cmd/delocate_wheel.py (hunk 1 above)
Comment (HexDecimal) [reply delocate/cmd/delocate_wheel.py]: The substring search by rossant is intentional.  The main alternative is to allow RegEx, which is usually overkill.

File: delocate/cmd/delocate_wheel.py (hunk 2 above)
Comment (HexDecimal) [reply delocate/cmd/delocate_wheel.py]: I've rewritten the code to use `action="append"` but I haven't tested it.

**Review by matthew-brett — COMMENTED** [review]
All looks good - but it still needs some tests - maybe in `delocate/tests/test_scripts.py` ?

**Comment by HexDecimal** [comment]
Maybe.  I wish the script running tests were using a standard tool rather than `ScriptRunner`.  I'll probably come back to this later.

**Comment by matthew-brett** [comment]
Yes, I wrote my own script testing framework around 2016, before there were others available - or at least - before I knew there were others available.

**Comment by HexDecimal** [comment]
Won't merge until tests are made which is a reasonable request.  The new tests can use `ScriptRunner` like the current tests, or `pytest-console-scripts` which is more standard.

**Comment by PhilipGarnero** [comment]
Can we merge this please ?
I know it's untested but it's still better than nothing. It's just one click away from being useful to people 🙇‍♂️

**Comment by PhilipGarnero** [comment]
I took a few minutes to write something that should do ok as a test :

```py
@pytest.mark.xfail(sys.platform != "darwin", reason="Needs macOS linkage.")
def test_fix_wheel_with_excluded_dylibs():
    with InTemporaryDirectory() as tmpdir:
        fixed_wheel, stray_lib = _fixed_wheel(tmpdir)
        _rename_module(fixed_wheel, "module.other", "test.whl")
        shutil.copyfile("test.whl", "test2.whl")
        # We exclude the stray library so it shouldn't be present in the wheel
        code, stdout, stderr = run_command(["delocate-wheel", "-e", "extfunc", "test.whl"])
        with InWheel("test.whl"): 
            dylibs = pjoin("plat_pkg", "fakepkg1", ".dylibs")
            assert_equal(os.listdir(dylibs), [])
        # We exclude a library that does not exist so we should have a normal behavior
        code, stdout, stderr = run_command(["delocate-wheel", "-e", "doesnotexist", "test2.whl"])
        _check_wheel("test2.whl", ".dylibs")
```

**Comment by HexDecimal** [comment]
I've added the test and cleaned up the commits. It could be better but now I think it's good enough to merge.

**Review by HexDecimal — APPROVED** [review]

**Comment by HexDecimal** [comment]
Moved some CLI logic to a shared module so that any new code can apply to both `delocate-wheel` and `delocate-path` without having to copy that code across both modules.

**Comment by webknjaz** [comment]
> Moved some CLI logic to a shared module so that any new code can apply to both `delocate-wheel` and `delocate-path` without having to copy that code across both modules.

Hey @HexDecimal, I think the changes in this PR may have resulted in breaking the delocate CLI extending a glob-style paths arg: https://github.com/matthew-brett/delocate/issues/71#issuecomment-1830984185.

**Pull request merged**